Add historical price endpoint

Kevinrwh opened the pull request.

Kevinrwh committed:
add historical prices endpoint

alex-power commented on src/main/java/org/galatea/starter/domain/IexHistoricalPrices.java:
```diff
@@ -0,0 +1,20 @@
+package org.galatea.starter.domain;
+
+import java.math.BigDecimal;
+import java.util.Date;
```
Is this the right Date object? How does the JSON look on the return value of this date?

alex-power replied:
Is this the right Date object? How does the JSON look on the return value of this date?

alex-power commented on src/test/resources/wiremock/mappings/mapping-historicalPriceEmptyRange.json:
```diff
@@ -0,0 +1,39 @@
+{
+  "name" : "historical_price_empty_range",
+  "request" : {
+    "url" : "/stock/chart/max/{date}?symbol=AAPL&range=",
```
What's this one for? I don't see a test that validates the behavior of passing an empty range.

Kevinrwh replied:
I had created a test for it but was having issues. I will look into doing it again and consult you if I have issues!

alex-power commented on src/test/resources/wiremock/mappings/mapping-historicalPriceEmptySymbol.json:
```diff
@@ -0,0 +1,33 @@
+  "response" : {
+    "status" : 200,
+    "jsonBody" : [{
+      "symbol": "Invalid"
```
Is this how the IEX api will actually respond when you pass symbol as blank? 200 with symbol: "Invalid"? Remember, wiremock should be matching the behavior of the IEX API when possible, so that the test is capturing actually possible scenarios when the app is running against the real IEX API.

Kevinrwh replied:
Yes, I was trying to generate an error message but will have to look into how the IEX API actually works

Kevinrwh committed:
add new error test and make response accurate to IEX API

Kevinrwh committed:
error path scenarios tested for historical prices

Kevinrwh committed:
added extra tests for error scenarios & corrected GET calls

Kevinrwh committed:
deleted uneccessary file

Kevinrwh committed:
add extra error scenarios and correct get mappings (actual)

Kevinrwh commented:
I deleted the 'Both Range and Date' mock since range is inclusive of date

Kevinrwh committed:
input validation and updated exception handling

Kevinrwh commented:
I updated feign exception to return different types of 4XX responses, 404 and bad request (400). I also added some input validation in the service logic to take care of null and empty strings

Kevinrwh committed:
revert to old state

Kevinrwh committed:
undo revert

Kevinrwh committed:
undo revert

Kevinrwh committed:
add correct endpoint using token and fix failing tests

alex-power commented on src/main/resources/application.yml:
```diff
@@ -47,6 +48,7 @@ spring:
    rest:
       # this points at the local WireMock server in the test environment.
       iexBasePath: http://localhost:${wiremock.server.port}/
+      iexToken: pk_8bf2fa8153b64102b1d34199273d2063/
```
don't commit secrets to code repos, this should remain in a local property file that is not included in the git

alex-power replied:
Dont' commit secrets to code repos, keep them in local config files that do not get committed

alex-power commented on src/main/java/org/galatea/starter/service/IexService.java:
```diff
@@ -45,5 +48,31 @@ public List<IexLastTradedPrice> getLastTradedPriceForSymbols(final List<String>
+        throw new HttpMessageNotReadableException("\"range\" is not allowed to be empty");
+      }
+      return iexClient.getHistoricalPricesRange(symbol, range);
+    } else if ((StringUtils.isBlank(range)) && StringUtils.isBlank(date) == false) { // Date specified but not range
```
I think you can do isNotBlank instead of isBlank == false

Kevinrwh replied:
done!

alex-power commented on src/main/java/org/galatea/starter/service/IexService.java:
```diff
@@ -45,5 +48,31 @@ public List<IexLastTradedPrice> getLastTradedPriceForSymbols(final List<String>
+    } else if(StringUtils.isBlank(range) && StringUtils.isBlank(date)){ // No range or date specified
+      return iexClient.getHistoricalPricesForSymbol(symbol);
+    } else if ((StringUtils.isBlank(range) == false) && StringUtils.isBlank(date)) { // Range specified but not date
+      if(range == ""){
```
isBlank covers the empty scneario for range, this is unnecessary

Kevinrwh replied:
I was having issues with isBlank for date and range since the IEX API response is different if range is an empty string vs when it's not included.

Kevinrwh replied:
made changes so that logic is old. The new commit does this correctly :)
